src/fw-logs/fw-logs-parser.cpp

```diff
@@ -1,90 +1,315 @@
+                }
+                structured->params_blob.assign( blob_start, blob_start + blob_size );
+            }
+        }
+
+        const fw_logs_formatting_options & fw_logs_parser::get_format_options_ref( int source_id ) const
+        {
+            auto iter = _source_to_formatting_options.find( source_id );
+            if( iter != _source_to_formatting_options.end() )
+                return iter->second;
 
-            return log_data;
+            throw librealsense::invalid_value_exception( rsutils::string::from()
+                                                         << "Invalid source ID received " << source_id );
+        }
+
+        std::string fw_logs_parser::get_source_name( int source_id ) const
+        {
+            auto iter = _source_id_to_name.find( source_id );
+            if( iter != _source_id_to_name.end() )
+                return iter->second;
+
+            throw librealsense::invalid_value_exception( rsutils::string::from()
+                                                         << "Invalid source ID received " << source_id );
+        }
+
+        rs2_log_severity fw_logs_parser::parse_severity( uint32_t severity ) const
+        {
+            return fw_logs::fw_logs_severity_to_rs2_log_severity( severity );
+        }
+
+        legacy_fw_logs_parser::legacy_fw_logs_parser( const std::string & xml_content )
+            : fw_logs_parser( xml_content )
+        {
+            // Check expected size here, no need to check repeatedly in parse functions
+            if( _source_to_formatting_options.size() != 1 )
+                throw librealsense::invalid_value_exception( rsutils::string::from()
+                                                             << "Legacy parser expect one formating options, have "
+                                                             << _source_to_formatting_options.size() );
+        }
+
+        size_t legacy_fw_logs_parser::get_log_size( const uint8_t * ) const
+        {
+            return sizeof( legacy_fw_log_binary );
+        }
+
+        void legacy_fw_logs_parser::structure_timestamp( const fw_logs_binary_data * raw,
+                                                         structured_binary_data * structured ) const
+        {
+            const auto actual_struct = reinterpret_cast< const legacy_fw_log_binary * >( raw->logs_buffer.data() );
+            structured->timestamp = actual_struct->timestamp;
+        }
+
+        void legacy_fw_logs_parser::structure_params( const fw_logs_binary_data * raw,
+                                                      size_t num_of_params,
+                                                      structured_binary_data * structured ) const
+        {
+            const auto actual_struct = reinterpret_cast< const legacy_fw_log_binary * >( raw->logs_buffer.data() );
+            uint16_t params_size_bytes = 0;
+            if( num_of_params > 0 )
```

Comment:
can the nuber of params change here? Isn't it always the same number? (4 params)?

Reply:
Not always same number of params.
Legacy format had placeholder for up to 3 parameters, but events can have less then that. We read the exact number for each event from the XML.